src/coreclr/jit/assertionprop.cpp

```diff
@@ -3416,19 +3437,33 @@ bool Compiler::optZeroObjAssertionProp(GenTree* tree, ASSERT_VALARG_TP assertion
         return false;
     }
 
-    unsigned       lclNum         = tree->AsLclVar()->GetLclNum();
+    const unsigned lclNum         = tree->AsLclVar()->GetLclNum();
     AssertionIndex assertionIndex = optLocalAssertionIsEqualOrNotEqual(O1K_LCLVAR, lclNum, O2K_ZEROOBJ, 0, assertions);
     if (assertionIndex == NO_ASSERTION_INDEX)
     {
         return false;
     }
 
+    // TODO: create proper simd zero constant
+    //
+    if (varTypeIsSIMD(tree))
+    {
+        return false;
+    }
+
     AssertionDsc* assertion = optGetAssertion(assertionIndex);
     JITDUMP("\nAssertion prop in " FMT_BB ":\n", compCurBB->bbNum);
     JITDUMPEXEC(optPrintAssertion(assertion, assertionIndex));
     DISPNODE(tree);
 
-    tree->BashToZeroConst(TYP_INT);
+    if (varTypeIsStruct(tree))
+    {
+        tree->BashToZeroConst(TYP_INT);
+    }
+    else
+    {
+        tree->BashToZeroConst(tree->TypeGet());
+    }
```

Comment:
Isn't the `!varTypeIsStruct` branch dead code?

(We have an assert to that effect at the top)

Reply:
Yep.